Feat/wizard permissions

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Makes the access checks for the wizards more versatile, not single-capability-based. This will allow access to various areas of the product for different users. 

### How to test the changes in this Pull Request:

1. Set `newspack-newsletters` to `feat/capabilities` branch (https://github.com/Automattic/newspack-newsletters/pull/1398)
2. Set `newspack-popups` to `feat/capabilities` branch (https://github.com/Automattic/newspack-popups/pull/1259)
1. As a site administrator, verify that you can see all the expected items in the Newspack Dashboard and the the wizards work as expected*
1. Install the [capability-manager-enhanced plugin](https://wordpress.org/plugins/capability-manager-enhanced/)
3. Create a new role in Capabilities -> Roles and then open Capabilities -> Capabilities. Choose the new role in the dropdown
4. Assign Posts -> Edit in "Editing" group and "read" in "Admin" group – this will let the user with the role access /wp-admin
5. Open the "Newspack" group and observe a list of capabilities – one for each wizard** 
6. Assign the `newspack wizard newspack-syndication-wizard` to the role and save it
7. Create a new user and assign the newly created role to them
8. Log in as the new user and observe the Newspack menu item feature the single chosen wizard, and likewise the Newspack Dashboard
9. One by one, test access to all wizards
    1. For Newsletters and Campaigns, apart from the wizard cap you'll also have to assign all caps present in the "Newspack Newsletters" and "Newspack Campaigns" groups

\* I know, that's a lot to ask. But frankly these changes touch everything about the wizards 😬
\** Note that `setup`, `settings`, and `components-demo` wizards won't appear on the dashboard. These are special wizards, not meant for access from dashboard:

- Setup wizard is used for a first-time Newspack setup
- Settings wizard does not have a UI, but it might at some point
- Components Demo is a showcase/development-help for the UI used in the Newspack admin areas. It can be accessed directly at `/wp-admin/admin.php?page=newspack-components-demo` or via a wizard footer link if `WP_NEWSPACK_DEBUG` env variable is set. 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

chickenn00dle commented:
Wizards:

- [ ] Reader Revenue
- [ ] Advertising
- [ ] Syndication
- [ ] Analytics
- [ ] SEO
- [ ] Engagement
- [ ] Campaigns

chickenn00dle reviewed (commented):
Hey @adekbadek, I tested this and ran into a few issues.

1. Commented about a fatal related to the Newsletters plugin
2. For some reason, despite the site design wizard being visible for my test role, I was still not able to access the site design menu:
![Screenshot 2024-01-19 at 16 07 27](https://github.com/Automattic/newspack-plugin/assets/17905991/4e348251-2029-4d17-bf33-5587f55e3008)
3. The support link appears to be associated with seemingly random capabilities. Here it is alongside the health check wizard capability:
![Screenshot 2024-01-19 at 16 23 12](https://github.com/Automattic/newspack-plugin/assets/17905991/43f166c2-e13a-4318-8be4-ebf1b863fca3)
And then again for the settings wizard and general wizard capabilities.
![Screenshot 2024-01-19 at 16 22 43](https://github.com/Automattic/newspack-plugin/assets/17905991/fbdcbfa7-1cb2-4ed7-98c9-ce9dbc0710d2)
I don't think it appeared for any others.
4. I wasn't sure what I needed to get the respective wizard for the `newspack wizard newspack-components-demo-wizard` capability to appear. I actually don't even have this on my admin account.
5. While going through the `Engagement` wizard ESP section for a new site, I noticed a fatal error appeared for the Subscription List section, and then that section would no longer load and I could not complete the ESP portion of the Engagement wizard. I wasn't able to get a screenshot of the error that appeared, but here is a screenshare of that section no longer appearing and me being unable to complete the ESP portion of the engagement wizard:

(ping me for screenshare)

I'm actually not sure if this is related to these changes or maybe the newsletters PR (or something else entirely)

adekbadek commented on includes/wizards/class-engagement-wizard.php:
```diff
@@ -465,4 +478,32 @@ public static function sanitize_categories( $categories ) {
+		}
+		update_user_meta( $current_user_id, $user_meta_name, true );
+
+		$cpt_caps = \Newspack_Newsletters::get_capabilities_list();
```
Done in https://github.com/Automattic/newspack-newsletters/pull/1398/commits/fd954c50f7735f54336761d37e44f05ea08616d3

adekbadek commented:
> 3. The support link appears to be associated with seemingly random capabilities.

This is added by a different plugin (an internal one). The dashboard items can be filtered and any plugin can add items – if the user has only the `newspack_wizard_newspack` capability, they'll see a blank dashboard or only the items added by other integrations. 

> 4. I wasn't sure what I needed to get the respective wizard for the newspack wizard `newspack-components-demo-wizard` capability to appear. I actually don't even have this on my admin account.

This is a special wizard, a demo of components. It'll be [available only in "debug mode"](https://github.com/Automattic/newspack-plugin/blob/1b65a73b532c5381a4d4cfa188848eaa36587ca1/includes/wizards/class-wizard.php#L122), that is with the `WP_NEWSPACK_DEBUG` env variable set. You can visit it directly at `/wp-admin/admin.php?page=newspack-components-demo`. I've updated the testing instructions now!

adekbadek commented:
> 2. For some reason, despite the site design wizard being visible for my test role, I was still not able to access the site design menu

Fixed in c72c1dfb1aa49184bdea1cb365e1c9dcc0633e4d – this wizard was slyly using REST endpoints of another wizard. 

> 5. While going through the Engagement wizard ESP section for a new site, I noticed a fatal error appeared for the Subscription List section […]

Fixed in d91e8885a8f7e14f1b3344223647caaa79d3b573

chickenn00dle reviewed (approved):
Thanks for clarifying @adekbadek 🙇 

I confirmed all of the issues I brought up have been resolved, so this is good to go 👍 

My only note would be we should probably hide any capabilities related to functionality hidden behind the newspack dev flag when the flag is not set. I am not sure how involved our publishers are when it comes to site admin, but it seems potentially confusing.

That said, its not really a blocker, so gonna give this a thumbs up.

adekbadek commented:
> we should probably hide any capabilities related to functionality hidden behind the newspack dev flag when the flag is not set.

I've approached is slightly differently – with a blacklist of capabilities which will not show up in the "Newspack" section of the CME plugin UI. These will still be available in the "Additional" section, but that's less likely to cause any confusion.
